File: LeetCode/src/combinationSum/CombinationSum2.java
package combinationSum;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSum2 {
	public List<List<Integer>> combinationSum(int[] candidates, int target) {
		List<List<Integer>> res = new ArrayList<>();// 结果
		List<Integer> temp = new ArrayList<>();
		Arrays.sort(candidates);// 先进行排序
		findComb(candidates, target, 0, 0, res, temp);
		return res;
	}

	private void findComb(int[] candidates, int taget, int level, int sum, List<List<Integer>> res,
			List<Integer> temp) {
		if (sum == taget) {
			res.add(new ArrayList<>(temp));
			return;
		} else if (sum > taget) {
			return;
		} else {
			for (int i = level; i < candidates.length; i++) {
				temp.add(candidates[i]);
				findComb(candidates, taget, i, sum + candidates[i], res, temp);
				temp.remove(temp.size() - 1);
			}
		}
	}
}

File: LeetCode/src/array/combinationSumII/Combination_Sum_2.cpp
#include <vector>	
#include <iostream>
#include <algorithm>

using namespace std;

//递归体
void getSum(vector<int>& candidates, int target, int pos, vector<int>& temp, vector<vector<int>>& res) {
	if (target < 0)
		return;
	if (target == 0) {
		res.push_back(vector<int>(temp));
		return;
	}
	int prev = -1;
	//DFS（也是俗称暴力解法）
	for (int i = pos; i < candidates.size(); ++i) {
		if (prev != candidates[i]) {
			temp.push_back(candidates[i]);
			getSum(candidates, target - candidates[i], i + 1, temp, res);
			temp.pop_back();
			prev = candidates[i];
		}
	}
}

vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
	sort(candidates.begin(), candidates.end());
	vector<vector<int>> res;
	getSum(candidates, target, 0, vector<int>(), res);
	return res;
}